Ensure usage of TraverseOperation is thread-safe

szewai opened the pull request:
#### b0571f4736c825f4e0bd44a4ced2fbe1835760e1
<pre>
Ensure usage of TraverseOperation is thread-safe
<a href="https://bugs.webkit.org/show_bug.cgi?id=283232">https://bugs.webkit.org/show_bug.cgi?id=283232</a>
<a href="https://rdar.apple.com/140041827">rdar://140041827</a>

Reviewed by Per Arne Vollan and Youenn Fablet.

TraverseOperation is accessed on different threads, see Storage::traverseWithinRootPath, but it is not thread-safe.
Its member `type` is a String and is accessed on different threads without protection (also it is not an isolated copy).
To fix this issue, this patch just removes the unnecessary data members from TraverseOperation, as they can be captured
in WorkQueue task instead. Also, this patch makes TraverseOperation class instead of struct, so access to data members
needs to go through member functions, which either have threading check or acquire needed lock. Finally, this patch
makes TraverseOperation ThreadSafeRefCounted so it could be captured in task instead of data members of Storage.

* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::~Storage):
(WebKit::NetworkCache::Storage::traverseWithinRootPath):
(WebKit::NetworkCache::Storage::TraverseOperation::TraverseOperation): Deleted.
(WebKit::NetworkCache::Storage::TraverseOperation::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:

Canonical link: <a href="https://commits.webkit.org/286751@main">https://commits.webkit.org/286751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/292d3cb2f5e693d8c4a74698ae3ac3fa2dd5c571

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29908 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/81556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28272 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65176 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4324 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60339 "32 flakes 105 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18411 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80060 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50290 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66094 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40651 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47692 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26598 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23920 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82976 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4373 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/2934 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68612 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4528 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66067 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67863 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16918 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11850 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4319 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7134 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7774 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6098 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -239,27 +239,57 @@ void Storage::WriteOperation::invokeCompletionHandler(int result)
+    }
+
+private:
+    TraverseOperation(Storage::TraverseHandler&& handler)
```
Explicit

pvollan reviewed (approved):
R=me.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -239,27 +239,57 @@ void Storage::WriteOperation::invokeCompletionHandler(int result)
-    Condition activeCondition;
-    unsigned activeCount WTF_GUARDED_BY_LOCK(activeLock) { 0 };
-};
+    static const unsigned maxParallelActivityCount = 5;
```
`constexpr`

basuke commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -239,27 +239,57 @@ void Storage::WriteOperation::invokeCompletionHandler(int result)
+        Locker locker { m_lock };
+        m_activeCondition.wait(m_lock, [this] {
+            assertIsHeld(m_lock);
+            return !m_activityCount;
```
I'm not sure how `waitUntilActivitiesFinished()` is called, but this seems potential problem:
1. thread A is active. `m_activityCount == 1` 
2. main thread is waiting by `waitUntilActivitiesFinished()`
3. thread A call `decrementActivityCount()`
4. during that, thread B call `waitAndIncrementActivityCount()`
5. thread A calls `notifyOne()` `m_activityCount == 0`
6. both main thread and thread B can be called, but if main thread is chosen, it satisfy the condition and goes on while thread B is still working.

szewai replied:
ah, the key is in existing usage, `waitAndIncrementActivityCount()` and `waitUntilActivitiesFinished()` is invoked in order on the same thread (in the same task dispatched to ` ioQueue()`). `decrementActivityCount()` is invoked on the main thread (IOChannel dispatches results back to main).

szewai committed:
Ensure usage of TraverseOperation is thread-safe
https://bugs.webkit.org/show_bug.cgi?id=283232
rdar://140041827

Reviewed by Per Arne Vollan and Youenn Fablet.

TraverseOperation is accessed on different threads, see Storage::traverseWithinRootPath, but it is not thread-safe.
Its member `type` is a String and is accessed on different threads without protection (also it is not an isolated copy).
To fix this issue, this patch just removes the unnecessary data members from TraverseOperation, as they can be captured
in WorkQueue task instead. Also, this patch makes TraverseOperation class instead of struct, so access to data members
needs to go through member functions, which either have threading check or acquire needed lock. Finally, this patch
makes TraverseOperation ThreadSafeRefCounted so it could be captured in task instead of data members of Storage.

* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::~Storage):
(WebKit::NetworkCache::Storage::traverseWithinRootPath):
(WebKit::NetworkCache::Storage::TraverseOperation::TraverseOperation): Deleted.
(WebKit::NetworkCache::Storage::TraverseOperation::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:

Canonical link: https://commits.webkit.org/286751@main
